### PR TITLE
WIP: Updates to Weaviate integration and documentation

### PR DESCRIPTION
**Pull request by bobvanluijt**
Goals:

- [x] Using Weaviate's native filtering for search
- [x] Using Weaviate cross-references to make relations in Documents
- [x] Using Weaviate native integration for meta filtering
- [ ] Updated documentation
- [x] Updated benchmark scripts
- [x] Added tests 
- [x] check and update documentation, if required. See [guide](https://github.com/jina-ai/jina/blob/master/CONTRIBUTING.md#documentation-guidelines)

### CODE REVIEW

**Comment by bobvanluijt** [comment]
Note, this is still WIP, please don't merge (yet)

**Comment by JoanFM** [comment]
Hello @bobvanluijt,

Thanks for the contribution. However, we will need to separate this contribution into separate PRs.

We would like to see a PR for each issue you are solving or feature implemented. The size of the PR will not be handlable, and also we would like to have more granularity on the changes accepted and not.

Thanks!

File: docarray/array/storage/weaviate/backend.py [hunk 1]
```diff
@@ -68,11 +68,38 @@ class WeaviateConfig:
...
+
+    Parameters
+    ----------
+    results : dict
+        The Weaviate batch creation return value.
+    """
+
+    if results is not None:
+        for result in results:
+            if 'result' in result and 'errors' in result['result']:
```
Comment (JoanFM) [inline docarray/array/storage/weaviate/backend.py:97]:
```suggestion
            if result.get('result', {}).get('errors', {}).get('error', None) is not None:
```

File: docarray/array/storage/weaviate/backend.py [hunk 2]
```diff
@@ -117,6 +144,14 @@ def _init_storage(
             f'{config.protocol}://{config.host}:{config.port}',
             timeout_config=config.timeout_config,
         )
+        self._client.batch(
```
Comment (JoanFM) [inline docarray/array/storage/weaviate/backend.py:147]: what does this batch do?

File: docarray/array/storage/weaviate/backend.py [hunk 3]
```diff
@@ -234,14 +327,10 @@ def _load_or_create_weaviate_schema(self):
...
         if not self._config.name:
-            name_candidate = self._get_weaviate_class_name()
-            doc_schemas = self._get_schema_by_name(name_candidate)
-            while self._client.schema.contains(doc_schemas):
-                name_candidate = self._get_weaviate_class_name()
-                doc_schemas = self._get_schema_by_name(name_candidate)
-            self._client.schema.create(doc_schemas)
-            self._config.name = name_candidate
-            return doc_schemas
+            raise ValueError(
```
Comment (JoanFM) [inline docarray/array/storage/weaviate/backend.py:330]: this is too much of a breaking change. It is a wanted feature to have a default name

File: docarray/array/storage/weaviate/find.py [hunk 4]
```diff
@@ -28,8 +29,65 @@
...
+    'offset',
+    'location']
+
+options_with_cross_refs = ['chunks', 'matches']
+
 
 class FindMixin:
+
+    def collect_results_from_weaviate(self, result, include_cosine):
+        """Returns a Doc based on Weaviate results
```
Comment (JoanFM) [inline docarray/array/storage/weaviate/find.py:53]: this shluld be private

File: docarray/array/storage/weaviate/getsetdel.py [hunk 5]
```diff
@@ -16,11 +16,19 @@ def _getitem(self, wid: str) -> 'Document':
...
+        try:   
+            resp = self._client.data_object.get_by_id(wid, class_name=self._config.name ,with_vector=True)
+
+            return_doc = Document(id=resp['properties']['_docarray_id'])
+
+            for prop in resp['properties']:            
+                if prop[:1] != '_':
+                    setattr(return_doc, prop, resp['properties'][prop])
+
+            if 'vector' in resp and resp['vector'] != None:
```
Comment (JoanFM) [inline docarray/array/storage/weaviate/getsetdel.py:28]: add a similar syntax as the one proposed before

File: scripts/benchmarking.py [hunk 6]
```diff
@@ -105,7 +105,7 @@ def recall(predicted, relevant, eval_at):
             {'n_dim': D},
         ),
         ('qdrant', {'n_dim': D, 'scroll_batch_size': 8}),
-        ('weaviate', {'n_dim': D}),
+        ('weaviate', {'n_dim': D, 'distance': 'l2-squared'}),
```
Comment (JoanFM) [inline scripts/benchmarking.py:108]: this changes should be in another PR

**Review by JoanFM — CHANGES_REQUESTED** [review]

**Comment by bobvanluijt** [comment]
Hi @JoanFM 👋 

> [...] we will need to separate this contribution into separate PRs.

That makes complete sense. Would it be an idea that we use this PR to complete everything, close it when done, and cut the PR up? It would be great if you can guide me through how you would like to see this ⛵

File: docarray/array/storage/weaviate/backend.py (hunk 3 above)
Comment (bobvanluijt) [reply docarray/array/storage/weaviate/backend.py:330]: This is the Weaviate client configuration when choosing Weaviate as a storage engine (hence it's in `_init_storage()`)

I've added additional information in the code itself

File: docarray/array/storage/weaviate/backend.py (hunk 2 above)
Comment (bobvanluijt) [reply docarray/array/storage/weaviate/backend.py:147]: This is the Weaviate client configuration when choosing Weaviate as a storage engine (hence it's in `_init_storage()`)

I've added additional information in the code itself.

File: docarray/array/storage/weaviate/backend.py (hunk 3 above)
Comment (bobvanluijt) [reply docarray/array/storage/weaviate/backend.py:330]: Very good point, this is changed

File: docarray/array/storage/weaviate/getsetdel.py (hunk 5 above)
Comment (bobvanluijt) [reply docarray/array/storage/weaviate/getsetdel.py:28]: Sorry, not 100% what you refer to here 🙈

File: scripts/benchmarking.py (hunk 6 above)
Comment (bobvanluijt) [reply scripts/benchmarking.py:108]: That makes sense, please see [this](https://github.com/jina-ai/docarray/pull/487#issuecomment-1214382690) and let me know if that works for you.

File: docarray/array/storage/weaviate/backend.py [hunk 7]
```diff
@@ -68,11 +69,38 @@ class WeaviateConfig:
...
+    ----------
+    results : dict
+        The Weaviate batch creation return value.
+    """
+
+    if results is not None:
+        for result in results:
+            if result.get('result', {}).get('errors', {}).get('error', None) is not None:
+                if 'error' in result['result']['errors']:
+                    raise ValueError(
```
Comment (JoanFM) [inline docarray/array/storage/weaviate/backend.py:100]: this extra if is not needed anymore

File: docarray/array/storage/weaviate/getsetdel.py (hunk 5 above)
Comment (JoanFM) [reply docarray/array/storage/weaviate/getsetdel.py:28]: I mean, use better the dict.get syntax proposed in another change

**Review by JoanFM — CHANGES_REQUESTED** [review]

**Comment by JoanFM** [comment]
> Hi @JoanFM 👋
> 
> > [...] we will need to separate this contribution into separate PRs.
> 
> That makes complete sense. Would it be an idea that we use this PR to complete everything, close it when done, and cut the PR up? It would be great if you can guide me through how you would like to see this ⛵

It works for me to complete and then divide, but then lets mark them as draft.

I will not be able to guide as I have low connectivity. I would suggest to separatr as much as possible in what issues u try to solve or the features u aim to add.

**Comment by bobvanluijt** [comment]
Ok, I'll finish it in draft and cut it up in pieces later if the tests are green 👍
